# MonArch: an automated pipeline to annotate circRNAs in prokaryotic RNA-Seq data

If you use the contents of this repository, please cite:
```
citation
```

# Usage
- detais on the command and input formats

# Instalation and requisites

We recommend using [Conda](https://docs.conda.io/en/latest/) for installing all MonArch dependencies. We provided a `.yml` file in the `misc` directory with all the dependencies. Go to the MonArch directory and create the enviroment with the dependencies with:
```
conda env create -f misc/monarch.yml
```
Before running MonArch, activate the enviroment with:
```
conda activate monarch
```

If you do not want to use Conda to install all the requisites (**not recommended**), below is a list of all the dependencies of MonArch. Be aware tat you might need to change how the programs are called in the main monarch.sh script since it assumes that the dependencies were installed with Conda.


# Outputs
- details of the outputs

# The Pipeline
- figure

# Known issues
- small genomes/data
- cannot effectvely use paired-end sequencing info

# License

[GNU General Public License, version 3](https://www.gnu.org/licenses/gpl-3.0.html)
